service/udp_linux.go

```diff
@@ -0,0 +1,122 @@
+// Copyright 2018 Jigsaw Operations LLC
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     https://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package service
+
+import (
+	"context"
+	"net"
+	"syscall"
+	"unsafe"
+
+	onet "github.com/Jigsaw-Code/outline-ss-server/net"
+	"golang.org/x/sys/unix"
+)
+
+func ListenUDP(network string, laddr *net.UDPAddr) (onet.UDPPacketConn, error) {
+	lc := &net.ListenConfig{
+		Control: func(network, address string, c syscall.RawConn) error {
+			return c.Control(func(fd uintptr) {
```

Comment:
Can you use https://pkg.go.dev/syscall#Bind or https://pkg.go.dev/syscall#BindToDevice at this point so you don't need to change the write code?

Reply:
> Or simply bind to a single interface.

How do you determine which interface to bind to? Most likely the user wants to bind to all interfaces.

> Can you use https://pkg.go.dev/syscall#Bind or https://pkg.go.dev/syscall#BindToDevice at this point so you don't need to change the write code?

`SO_BINDTODEVICE` does not allow you to specify a source IP. If an interface has multiple IP addresses of the same address family, outbound packets are always sent via the default IP. `IP_PKTINFO` allows you to set a source IP in addition to an interface index. With the source IP setting, you don't have to bind the socket to an IP to send from that IP.

> Why not?

For every interface you would have to create one `clientConn`. My implementation creates one `clientConn` that listens on all interfaces and IPs. This is also how `wireguard-go` handles multiple interfaces.